### PR TITLE
[opt](file_reader) add prefetch buffer to read csv&json file

### PR DESCRIPTION
**Pull request by AshinGau**
# Proposed changes
Co-authored-by: ByteYue <[yj976240184@gmail.com](mailto:yj976240184@gmail.com)>
This PR is an optimization for https://github.com/apache/doris/pull/17478:
1. Change the buffer size of `LineReader` to 4MB to align with the size of prefetch buffer.
2. Lazily prefetch data in the first read to prevent wasted reading.
3. S3 block size is 32MB only, which is too small for a file split. Set 128MB as default file split size.
4. Add `_end_offset` for prefetch buffer to prevent wasted reading.

The query performance of reading data on object storage is improved by more than 3x+.

### S3 read optimization
**Throughput bottleneck in S3**
**Single threaded reads**: Data is read synchronously on a single thread which results in jobs spending most of the time waiting for data to be read over the network.

So add prefetch buffers to read sequential file(csv&json). Parquet&ORC will be optimized in next PR.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by AshinGau** [comment]
run buildall

File: be/src/io/fs/hdfs_file_reader.cpp [hunk 1]
```diff
@@ -40,6 +40,10 @@ HdfsFileReader::~HdfsFileReader() {
 }
 
 Status HdfsFileReader::close() {
+    if (_closed) {
+        return Status::OK();
+    }
+    std::lock_guard<std::mutex> lock(_lock);
```
Comment (morningman) [inline be/src/io/fs/hdfs_file_reader.cpp:46]: Is it necessary to make a hdfs reader thread safe?

**Comment by AshinGau** [comment]
Co-authored-by: ByteYue <[yj976240184@gmail.com](mailto:yj976240184@gmail.com)>

File: be/src/common/config.h [hunk 2]
```diff
@@ -892,6 +892,9 @@ CONF_mInt16(pipeline_short_query_timeout_s, "20");
 // Will remove after fully test.
 CONF_Bool(enable_index_apply_preds_except_leafnode_of_andnode, "true");
 
+// buffered prefetch
+CONF_mInt32(prefetch_single_buffer_size_mb, "4");
```
Comment (morningman) [inline be/src/common/config.h:896]: I think we can remove this config, use a fixed value, if there is no need to modify this.

File: be/src/io/fs/buffered_reader.h [hunk 3]
```diff
@@ -29,6 +31,107 @@
...
+                        const IOContext* io_ctx) override;
+
+private:
+    size_t get_buffer_pos(int64_t position) const {
+        return (position % _whole_pre_buffer_size) / s_max_pre_buffer_size;
+    }
+    size_t get_buffer_offset(int64_t position) const {
+        return (position / s_max_pre_buffer_size) * s_max_pre_buffer_size;
+    }
+    void resetAllBuffer(size_t position) {
```
Comment (morningman) [inline be/src/io/fs/buffered_reader.h:115]:
```suggestion
    void reset_all_buffer(size_t position) {
```

File: be/src/io/fs/buffered_reader.cpp [hunk 4]
```diff
@@ -28,6 +28,166 @@
...
+BufferedReader::BufferedReader(io::FileReaderSPtr reader, int64_t offset, int64_t length,
+                               int64_t buffer_size)
+        : _reader(std::move(reader)), _start_offset(offset), _end_offset(offset + length) {
+    if (buffer_size == -1L) {
+        buffer_size = config::remote_storage_read_buffer_mb * 1024 * 1024;
+    }
+    _size = _reader->size();
+    _whole_pre_buffer_size = buffer_size;
+#ifdef BE_TEST
+    s_max_pre_buffer_size = config::prefetch_single_buffer_size_mb;
```
Comment (morningman) [inline be/src/io/fs/buffered_reader.cpp:140]:
```suggestion
    s_max_pre_buffer_size = config::prefetch_single_buffer_size_mb * 1024 * 1024;
```

File: be/src/io/fs/buffered_reader.cpp [hunk 5]
```diff
@@ -28,6 +28,166 @@
...
+    _buffer_status = BufferStatus::PREFETCHED;
+    _prefetched.notify_all();
+    // eof would come up with len == 0, it would be handled by read_buffer
+}
+
+Status PrefetchBuffer::read_buffer(size_t off, const char* out, size_t buf_len,
+                                   size_t* bytes_read) {
+    if (UNLIKELY(off >= _end_offset)) {
+        // Reader can read out of [_start_offset, _end_offset) by synchronous method.
+        IOContext io_context;
```
Comment (morningman) [inline be/src/io/fs/buffered_reader.cpp:88]:
```suggestion
```

File: be/src/io/fs/buffered_reader.cpp [hunk 6]
```diff
@@ -28,6 +28,166 @@
...
+        }
+        _buffer_status = BufferStatus::PENDING;
+        _prefetched.notify_all();
+    }
+    _len = 0;
+    Status s;
+    IOContext io_context;
+
+    size_t buf_size = _end_offset - _offset > _size ? _size : _end_offset - _offset;
+    s = _reader->read_at(_offset, Slice {_buf.data(), buf_size}, &_len, &io_context);
```
Comment (morningman) [inline be/src/io/fs/buffered_reader.cpp:73]:
```suggestion
    s = _reader->read_at(_offset, Slice {_buf.data(), buf_size}, &_len);
```

File: be/src/io/fs/buffered_reader.cpp [hunk 7]
```diff
@@ -28,6 +28,166 @@
...
+        if (UNLIKELY(_buffer_status == BufferStatus::CLOSED)) {
+            _prefetched.notify_all();
+            return;
+        }
+        _buffer_status = BufferStatus::PENDING;
+        _prefetched.notify_all();
+    }
+    _len = 0;
+    Status s;
+    IOContext io_context;
```
Comment (morningman) [inline be/src/io/fs/buffered_reader.cpp:70]:
```suggestion
```

File: be/src/io/fs/hdfs_file_reader.cpp (hunk 1 above)
Comment (AshinGau) [reply be/src/io/fs/hdfs_file_reader.cpp:46]: The reader inner `BufferedReader` should be thread safe, otherwise it will core dump.

**Comment by AshinGau** [comment]
run buildall

**Comment by AshinGau** [comment]
run buildall

**Comment by morningman** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.91 seconds
 stream load tsv:          425 seconds loaded 74807831229 Bytes, about 167 MB/s
 stream load json:         24 seconds loaded 2358488459 Bytes, about 93 MB/s
 stream load orc:          72 seconds loaded 1101869774 Bytes, about 14 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230404084947_clickbench_pr_124577.html

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**